### PR TITLE
[FLINK-18630] Redo the LongRideAlerts solutions to cover a corner case

### PR DESCRIPTION
**Pull request by alpinegizmo**
Fixing an issue in the reference solutions for this exercise. I've added a test to cover this corner case, which is when the END event for a ride arrives more than two hours before the matching START event. Thanks to @zbyszkop for pointing out the problem, and proposing a solution.

I also added some content to the DISCUSSION page (which was empty).

### CODE REVIEW

File: long-ride-alerts/DISCUSSION.md [hunk 1]
```diff
@@ -21,7 +21,23 @@ under the License.
 
 (Discussion of [Lab: `ProcessFunction` and Timers (Long Ride Alerts)](./))
 
+It would be interesting to test that the solution does not leak state.
 
+A good way to write unit tests for a `KeyedProcessFunction` that check for state retention, etc., is to
```
Comment (NicoK) [inline long-ride-alerts/DISCUSSION.md]:
```suggestion
A good way to write unit tests for a `KeyedProcessFunction` to check for state retention, etc., is to
```

File: long-ride-alerts/DISCUSSION.md [hunk 2]
```diff
@@ -21,7 +21,23 @@ under the License.
...
+[documentation on testing](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/testing.html#unit-testing-stateful-or-timely-udfs--custom-operators). 
+
+In fact, the reference solutions will leak state in the case where a START event is missing. They also
+leak in the case where the alert is generated, but then the END event does eventually arrive (after `onTimer()`
+has cleared the matching START event).
+
+This could be addressed either by using [state TTL](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/state.html#state-time-to-live-ttl),
+or by using another timer that eventually
+clears any remaining state. There is a tradeoff here, however: once that state has been removed,
+then if the matching events are't actually missing, but are instead very, very late, they will cause erroneous alerts.
```
Comment (NicoK) [inline long-ride-alerts/DISCUSSION.md]:
```suggestion
then if the matching events are not actually missing, but are instead very, very late, they will cause erroneous alerts.
```

File: long-ride-alerts/README.md [hunk 3]
```diff
@@ -62,13 +62,11 @@ The resulting stream should be printed to standard out.
 <details>
 <summary><strong>Overall approach</strong></summary>
 
-This exercise revolves around using a `ProcessFunction` to manage some keyed state and event time timers, and doing so in a way that works even when the END event for a given `rideId` arrives before the START (which will happen). The challenge is figuring out what state to keep, and when to set and clear that state.
-</details>
-
-<details>
-<summary><strong>Timers and State</strong></summary>
-
-You will want to use event time timers that fire two hours after the incoming events, and in the `onTimer()` method, collect START events to the output only if a matching END event hasn't yet arrived. As for what state to keep, it is enough to remember the "last" event for each `rideId`, where "last" is based on event time and ride type (START vs END &mdash; yes, there are rides where the START and END have the same timestamp), rather than the order in which the events are processed. The `TaxiRide` class implements `Comparable`; feel free to take advantage of that, and be sure to eventually clear any state you create.
```
Comment (NicoK) [inline long-ride-alerts/README.md:71]: Why did you remove the hints on what state to keep?

**Review by NicoK — CHANGES_REQUESTED** [review]
Thanks for this fix. At first, I stumbled upon the (intentional) leak that was handled differently in the previous version, but let's keep it that way to spark a discussion.

I left a few comments inline.

File: long-ride-alerts/README.md (hunk 3 above)
Comment (alpinegizmo) [reply long-ride-alerts/README.md:71]: That hint was somewhat wrong-headed. I've written a new hint that points (rather strongly) toward the new solution.

File: long-ride-alerts/README.md [hunk 4]
```diff
@@ -69,6 +69,20 @@ You will want to use event time timers that fire two hours after an incoming STA
 collect START events to the output only if a matching END event hasn't yet arrived.
 </details>
 
+<details>
+<summary><strong>State and timers</strong></summary>
+
+There are many possible solutions for this exercise, but in general it is enough to keep one
+`TaxiRide` in state (one `TaxiRide` for each key, or `rideId`). The approach used in the reference solution is to
+store whichever event arrives first (the START or the END), and if it's a START event,
+create a timer for two hours later. If and when the other event (for the same rideId) arrives,
```
Comment (NicoK) [inline long-ride-alerts/README.md]:
```suggestion
create a timer for two hours later. If and when the other event (for the same `rideId`) arrives,
```

File: long-ride-alerts/README.md [hunk 5]
```diff
@@ -69,6 +69,20 @@ You will want to use event time timers that fire two hours after an incoming STA
...
+carefully clean things up.
+
+It's possible to arrange this so that if `onTimer()` is called, you are guaranteed that
```
Comment (NicoK) [inline long-ride-alerts/README.md]:
```suggestion
It is possible to arrange this so that if `onTimer()` is called, you are guaranteed that
```

**Review by NicoK — CHANGES_REQUESTED** [review]

**Review by NicoK — APPROVED** [review]
Thanks, I'm going to approve the PR but left two very minor things to update if you like to.
